uniffi/src/lib.rs

```diff
@@ -258,415 +225,83 @@ pub fn check_remaining(buf: &[u8], num_bytes: usize) -> Result<()> {
-    // important that it consume ownership of the String, which is relinquished to the
-    // foreign language code (and can be restored by it passing the pointer back).
-    fn lower(obj: Self::RustType) -> Self::FfiType {
-        RustBuffer::from_vec(obj.into_bytes())
-    }
-
-    // The argument here *must* be a uniquely-owned `RustBuffer` previously obtained
-    // from `lower` above, and hence must be the bytes of a valid rust string.
-    fn try_lift(v: Self::FfiType) -> Result<Self::RustType> {
-        let v = v.destroy_into_vec();
-        // This turns the buffer back into a `String` without copying the data
-        // and without re-checking it for validity of the utf8. If the `RustBuffer`
-        // came from a valid String then there's no point in re-checking the utf8,
-        // and if it didn't then bad things are probably going to happen regardless
-        // of whether we check for valid utf8 data or not.
-        Ok(unsafe { String::from_utf8_unchecked(v) })
-    }
-
-    fn write(obj: Self::RustType, buf: &mut Vec<u8>) {
-        // N.B. `len()` gives us the length in bytes, not in chars or graphemes.
-        // TODO: it would be nice not to panic here.
-        let len = i32::try_from(obj.len()).unwrap();
-        buf.put_i32(len); // We limit strings to u32::MAX bytes
-        buf.put(obj.as_bytes());
-    }
-
-    fn try_read(buf: &mut &[u8]) -> Result<Self::RustType> {
-        check_remaining(buf, 4)?;
-        let len = usize::try_from(buf.get_i32())?;
-        check_remaining(buf, len)?;
-        // N.B: In the general case `Buf::chunk()` may return partial data.
-        // But in the specific case of `<&[u8] as Buf>` it returns the full slice,
-        // so there is no risk of having less than `len` bytes available here.
-        let bytes = &buf.chunk()[..len];
-        let res = String::from_utf8(bytes.to_vec())?;
-        buf.advance(len);
-        Ok(res)
+/// Helper function to lower an `anyhow::Error` that's wrapping an error type
+pub fn lower_anyhow_error_or_panic<FC, E>(
+    err: anyhow::Error,
+    arg_name: &str,
+) -> <FC as FfiConverter<E>>::FfiType
+where
+    FC: FfiConverter<E>,
+    E: 'static + Sync + Send + std::fmt::Debug + std::fmt::Display,
+{
+    match err.downcast::<E>() {
+        Ok(actual_error) => FC::lower(actual_error),
+        Err(ohno) => panic!("Failed to convert arg '{arg_name}': {ohno}"),
     }
 }
 
-/// A helper trait to implement lowering/lifting using a `RustBuffer`
+/// Macro to implement lowering/lifting using a `RustBuffer`
 ///
 /// For complex types where it's too fiddly or too unsafe to convert them into a special-purpose
 /// C-compatible value, you can use this trait to implement `lower()` in terms of `write()` and
 /// `lift` in terms of `read()`.
-pub trait RustBufferFfiConverter: Sized {
-    type RustType;
-    fn write(obj: Self::RustType, buf: &mut Vec<u8>);
-    fn try_read(buf: &mut &[u8]) -> Result<Self::RustType>;
-}
```

Comment:
This trait and the blanket impl below was annoying me, because small changes could lead to overlapping definitions.  I changed it to use a macro istead.